internal/pkg/override/patch.go

```diff
@@ -0,0 +1,318 @@
+// It supports a single file (cfn.patches.yml) with configured patches.
+func WithPatch(filePath string, opts PatchOpts) *Patch {
+	fs := afero.NewOsFs()
+	if opts.FS != nil {
+		fs = opts.FS
+	}
+
+	return &Patch{
+		filePath: filePath,
+		fs:       fs,
+	}
+}
+
+// Override returns the overriden CloudFormation template body
+// after applying YAML patches to it.
+func (p *Patch) Override(body []byte) ([]byte, error) {
+	patches, err := unmarshalPatches(p.filePath, p.fs)
+	if err != nil {
+		return nil, err
+	}
+
+	var root yaml.Node
+	if err := yaml.Unmarshal(body, &root); err != nil {
+		return nil, fmt.Errorf("invalid template: %w", err)
+	}
+
+	for i, patch := range patches {
+		var err error
+		switch patch.Operation {
+		case "add":
+			err = patch.applyAdd(&root)
+		case "remove":
+			err = patch.applyRemove(&root)
+		case "replace":
+			err = patch.applyReplace(&root)
+		default:
+			return nil, fmt.Errorf("unsupported operation %q: supported operations are %q, %q, and %q.", patch.Operation, "add", "remove", "replace")
+		}
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply the %q patch at index %d: %w", patch.Operation, i, err)
+		}
+	}
+
+	out, err := yaml.Marshal(&root)
+	if err != nil {
+		return nil, fmt.Errorf("unable to return modified document to []byte: %w", err)
+	}
+	return out, nil
+}
+
+func unmarshalPatches(path string, fs afero.Fs) ([]yamlPatch, error) {
+	content, err := afero.ReadFile(fs, path)
+	if err != nil {
+		return nil, fmt.Errorf("read file at %q: %w", path, err)
+	}
+
+	var patches []yamlPatch
+	if err := yaml.Unmarshal(content, &patches); err != nil {
+		return nil, fmt.Errorf("file at %q does not conform to the YAML patch document schema: %w", path, err)
+	}
+
+	return patches, nil
+}
+
+type yamlPatch struct {
+	Operation string `yaml:"op"`
+
+	// Path is in JSON Pointer syntax: https://www.rfc-editor.org/rfc/rfc6901
+	Path  string    `yaml:"path"`
+	Value yaml.Node `yaml:"value"`
+}
+
+func (p *yamlPatch) applyAdd(root *yaml.Node) error {
+	if p.Value.IsZero() {
+		return fmt.Errorf("value required")
+	}
+
+	pointer := p.Pointer()
+	parent, err := getNode(root, pointer.Parent(), nil)
+	if err != nil {
+		return err
+	}
+
+	switch parent.Kind {
+	case yaml.DocumentNode:
+		return parent.Encode(p.Value)
+	case yaml.MappingNode:
+		i, err := findInMap(parent, pointer.FinalKey(), pointer.Parent())
+		if err == nil {
+			// if the key is in this map, they are trying to replace it
+			return parent.Content[i+1].Encode(p.Value)
+		}
+
+		// if the key isn't in this map, then we need to create it for them
+		parent.Content = append(parent.Content, &yaml.Node{
+			Kind:  yaml.ScalarNode,
+			Tag:   "!!str",
+			Value: pointer.FinalKey(),
+		})
+		parent.Content = append(parent.Content, &p.Value)
+	case yaml.SequenceNode:
+		if pointer.FinalKey() == "-" {
+			// add to end of sequence
+			parent.Content = append(parent.Content, &p.Value)
+			return nil
```

Comment:
Is this possible to make this whole thing `parent.add(p.Value)`? sorry i can't include more lines but it should reach to the end.

and `applyAdd(root *yaml.Node)` will be clearer: we find the target node to add/replace/delete, and then apply the action.

Reply:
problem is that replace doesn't use the parent node, it uses the "node" node - so depending on the action, we need a different node

Reply:
though i guess the action for replace could be - take the parent, get the correct child node using `getNode()` and replace that?

Reply:
no i just meant we can wrap this up to make the logic clearer. It can be different for replace doesn't necessarily need to be `parent.`

Reply:
i think all of the logic for each operation (add/remove/replace) is totally separate, so i'm not sure there's much benefit in breaking out the find parent and "do operation" logic, since it can't get reused in any way...